### PR TITLE
Feat/User Profiles

### PR DESCRIPTION
**Pull request by naumovski-filip**
## Description

Testing config: **maqet3qh**

### TODO:
- ~~Move Profiles logic from account service and controller to Profiles service and controller.~~
- ~~Move edit, delete and create profile handlers to hook~~
- Unit and E2E tests

### Steps completed:

According to our definition of done, I have completed the following steps:

- [X] Acceptance criteria met
- [ ] Unit tests added
- [ ] Docs updated (including config and env variables)
- [X] Translations added
- [X] UX tested
- [X] Browsers / platforms tested
- [X] Rebased & ready to merge without conflicts
- [X] Reviewed own code

### CODE REVIEW

File: src/components/ProfileBox/AddNewProfile.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import Plus from '#src/icons/Plus';
+
+type Props = {
+  onClick: () => void;
+};
+
+const AddNewProfile = ({ onClick }: Props) => {
+  const { t } = useTranslation('user');
+  return (
+    <div onClick={onClick} className={classNames(styles.wrapper, styles.addProfileContainer)}>
```
Comment (ChristiaanScheermeijer) [inline src/components/ProfileBox/AddNewProfile.tsx]: When using `<div>` as button, you should also make it keyboard accessible:

```suggestion
  const { t } = useTranslation('user');
  const keyDownHandler = (event: React.MouseEvent<HTMLDivElement>) => (event.key === 'Enter' || event.key === ' ') && onClick()

  return (
    <div onClick={onClick} tabIndex="0" onKeyDown={keyDownHandler} className={classNames(styles.wrapper, styles.addProfileContainer)}>
```

File: src/components/ProfileBox/ProfileBox.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+const ProfileBox = ({ name, image, adult = true, editMode = false, onClick, onEdit, selected = false }: Props) => {
+  return (
+    <div className={styles.wrapper}>
+      <div className={classNames(styles.inner, selected && styles.selected)}>
+        <div onClick={onClick} className={styles.box}>
+          <img className={styles.image} src={image} alt="" />
+          {!adult && <span className={styles.kidsLabel}>Kids</span>}
+        </div>
+        {editMode && (
+          <div onClick={onEdit} className={styles.overlay}>
```
Comment (ChristiaanScheermeijer) [inline src/components/ProfileBox/ProfileBox.tsx]: Here also, `tabIndex` and `onKeyDown` props are needed (or use Button/IconButton instead)

File: src/components/UserMenu/UserMenu.tsx [hunk 3]
```diff
@@ -21,6 +27,18 @@ type Props = {
...
+  const { accessModel } = useConfigStore();
+  const { profile: currentProfile } = useProfileStore();
+  const profilesEnabled = useProfilesFeatureEnabled();
+
+  const { data, isFetching } = useListProfiles();
+  const profiles = data?.responseData.collection;
+
+  if (profilesEnabled && !profiles?.length) {
+    unpersistProfile();
+  }
```
Comment (ChristiaanScheermeijer) [inline src/components/UserMenu/UserMenu.tsx]: This logic should be elsewhere (no stores in components). The `unpersistProfile` seems a bit out of place, but that can be just me 😅

File: src/components/UserMenu/UserMenu.tsx [hunk 4]
```diff
@@ -33,9 +51,52 @@ const UserMenu = ({ showPaymentsItem, small = false, onClick }: Props) => {
...
   return (
     <ul className={styles.menuItems}>
+      {accessModel === 'SVOD' && profilesEnabled && (
+        <>
+          <div className={styles.sectionHeader}>{t('nav.switch_profiles')}</div>
+          {selectProfile.isLoading || isFetching ? (
+            <LoadingOverlay inline />
+          ) : (
+            profiles?.map((profile) => (
+              <li key={profile.id}>
```
Comment (ChristiaanScheermeijer) [inline src/components/UserMenu/UserMenu.tsx]: This can result in faulty HTML `ul > div + li`

File: src/containers/AppRoutes/AppRoutes.tsx [hunk 5]
```diff
@@ -12,12 +13,33 @@ import LegacySeries from '#src/pages/LegacySeries/LegacySeries';
...
+  const { accessModel } = useConfigStore(({ config, accessModel }) => ({ config, accessModel }), shallow);
+  const user = useAccountStore(({ user }) => user, shallow);
+  const { profile } = useProfileStore();
+  const profilesEnabled = useProfilesFeatureEnabled();
+
+  const shouldManageProfiles = !!user && profilesEnabled && !profile && (accessModel === 'SVOD' || accessModel === 'AUTHVOD') && !userModal;
 
   return (
     <Routes>
+      <Route index={shouldManageProfiles} element={<Navigate to="/u/profiles" />} />
```
Comment (ChristiaanScheermeijer) [inline src/containers/AppRoutes/AppRoutes.tsx]: This only works when the user navigates to the root `/`. Should this also be working for deep links?

A solution to this would be to directly redirect and pass a `?goto` query param so that the user ends up on the expected page after choosing a profile.

File: src/containers/Profiles/Profiles.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+  useEffect(() => {
+    if (!profilesEnabled || !user?.id || isError) navigate('/');
+  }, [profilesEnabled, navigate, user?.id, isError]);
+
+  const selectProfile = useSelectProfile();
+
+  useEffect(() => {
+    refetch();
+  }, [refetch]);
```
Comment (ChristiaanScheermeijer) [inline src/containers/Profiles/Profiles.tsx]: Why is this needed? 🤔

File: src/containers/Profiles/Profiles.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+  const breakpoint: Breakpoint = useBreakpoint();
+  const isMobile = breakpoint === Breakpoint.xs;
+
+  const { data, isLoading, isFetching, refetch, isError } = useListProfiles();
+  const activeProfiles = data?.responseData.collection.length || 0;
+  const canAddNew = activeProfiles < MAX_PROFILES;
+
+  useEffect(() => {
+    if (!profilesEnabled || !user?.id || isError) navigate('/');
```
Comment (ChristiaanScheermeijer) [inline src/containers/Profiles/Profiles.tsx]: Will this not create a redirect loop because the index route redirects back to the profiles page?

**Review by ChristiaanScheermeijer — COMMENTED** [review]
This is already much cleaner! I have left some minor feedback. I still have to try it out though

File: src/containers/Profiles/Profiles.tsx (hunk 6 above)
Comment (naumovski-filip) [reply src/containers/Profiles/Profiles.tsx]: To refetch on mount, for cases such as logout into login with a new account

**Comment by naumovski-filip** [comment]
@ChristiaanScheermeijer I've taken a look at the comments and made some changes accordingly

File: src/containers/Profiles/Profiles.tsx (hunk 6 above)
Comment (ChristiaanScheermeijer) [reply src/containers/Profiles/Profiles.tsx]: In that case, you can better do either;

a) Enable `refetchOnMount` in the [useQuery](https://tanstack.com/query/v3/docs/react/reference/useQuery)
b) Call `queryClient.invalidateQueries(['listProfiles'])` when the user logs out

The current code will always result in unnecessary/duplicate requests to the list profiles endpoint.

File: src/containers/Layout/Layout.tsx [hunk 8]
```diff
@@ -43,7 +44,9 @@ const Layout = () => {
     shallow,
   );
   const { updateSearchQuery, resetSearchQuery } = useSearchQueryUpdater();
-  const isLoggedIn = !!useAccountStore((state) => state.user);
+  const user = useAccountStore(({ user }) => user, shallow);
```
Comment (ChristiaanScheermeijer) [inline src/containers/Layout/Layout.tsx]: There is no need to shallow-check the `user` property because it should be immutable. 

```suggestion
  const user = useAccountStore(({ user }) => user);
```

File: src/containers/Profiles/DeleteProfile.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      <Dialog open={!!viewParam} onClose={closeHandler}>
+        {isDeleting && <LoadingOverlay inline />}
+        <div className={styles.deleteModal}>
+          <div>
+            <h2>{t('profile.delete')}</h2>
+            <p>{t('profile.delete_confirmation')}</p>
+          </div>
+          <div className={styles.deleteButtons}>
+            <Button label="Delete" color="delete" variant="contained" onClick={deleteHandler} />
+            <Button label="Cancel" variant="outlined" onClick={closeHandler} />
```
Comment (ChristiaanScheermeijer) [inline src/containers/Profiles/DeleteProfile.tsx]: Missing translations?

File: src/containers/Profiles/EditProfile.tsx [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            value: selectedAvatar || '',
+          }}
+          showCancelButton={!contained}
+          isMobile={isMobile}
+        />
+        <div className={styles.panel}>
+          <div className={styles.panelHeader}>
+            <h3>{t('profile.delete')}</h3>
+          </div>
+          <div className={profileStyles.profileInfo}>{t(`profile.delete_${profileDetails?.default ? 'main' : 'description'}`)}</div>
```
Comment (ChristiaanScheermeijer) [inline src/containers/Profiles/EditProfile.tsx]: We should ensure that translations are extractable by i18next-parser. This means the keys can't be template literals.

```suggestion
 <div className={profileStyles.profileInfo}>{profileDetails?.default ? t('profile.delete_main') : t('profile.delete_description')}</div>
```

File: package.json [hunk 11]
```diff
@@ -32,7 +32,7 @@
     "lint:stylelint": "stylelint '**/*.{css,scss}'",
     "commit-msg": "commitlint --edit $1",
     "codecept:mobile": "cd test-e2e && rm -rf \"./output/mobile\" && codeceptjs --config ./codecept.mobile.js run-workers --suites ${WORKER_COUNT:=8} ",
-    "codecept:desktop": "cd test-e2e && rm -rf \"./output/desktop\" && codeceptjs --config ./codecept.desktop.js run-workers --suites ${WORKER_COUNT:=8} ",
+    "codecept:desktop": "cd test-e2e && rm -rf \"./output/desktop\" && codeceptjs --config ./codecept.desktop.js run-workers --suites ${WORKER_COUNT:=2} ",
```
Comment (kmitrovv) [inline package.json]: revert workers

File: src/components/UserMenu/UserMenu.tsx [hunk 12]
```diff
@@ -33,9 +46,52 @@ const UserMenu = ({ showPaymentsItem, small = false, onClick }: Props) => {
 
   return (
     <ul className={styles.menuItems}>
+      {accessModel === 'SVOD' && profilesEnabled && (
```
Comment (kmitrovv) [inline src/components/UserMenu/UserMenu.tsx:48]: What do you think of wrapping this in another component, e.g UserMenuProfiles, ProfilesMenu, that can still be under UserMenu dir.

File: src/containers/Layout/Layout.tsx [hunk 13]
```diff
@@ -19,12 +19,17 @@ import MenuButton from '#components/MenuButton/MenuButton';
...
+import { unpersistProfile, useListProfiles, useProfilesFeatureEnabled } from '#src/hooks/useProfiles';
 
 const Layout = () => {
   const location = useLocation();
   const navigate = useNavigate();
   const { t, i18n } = useTranslation('common');
   const { config, accessModel } = useConfigStore(({ config, accessModel }) => ({ config, accessModel }), shallow);
+  const { data: profilesData } = useListProfiles();
+  const profiles = profilesData?.responseData.collection;
+  const profilesEnabled = useProfilesFeatureEnabled();
```
Comment (kmitrovv) [inline src/containers/Layout/Layout.tsx]: Why do we need to fetch profiles and  check profilesEnabled data separately in various parts of our codebase? What do you think of centralizing the logic to handle profilesData queries based on the profilesEnabled?

File: src/containers/Layout/Layout.tsx [hunk 14]
```diff
@@ -43,7 +52,9 @@ const Layout = () => {
     shallow,
   );
   const { updateSearchQuery, resetSearchQuery } = useSearchQueryUpdater();
-  const isLoggedIn = !!useAccountStore((state) => state.user);
+  const user = useAccountStore(({ user }) => user);
```
Comment (kmitrovv) [inline src/containers/Layout/Layout.tsx]: I guess this can be one liner as previously.. not seeing any benefit of splitting it

File: src/services/inplayer.service.ts [hunk 15]
```diff
@@ -0,0 +1,35 @@
+export const getBaseUrl = (sandbox: boolean) => (sandbox ? 'https://services-daily.inplayer.com' : 'https://services.inplayer.com');
```
Comment (kmitrovv) [inline src/services/inplayer.service.ts]: If all the methods are updated to be used from the sdk, this file is no longer needed.

**Review by kmitrovv — APPROVED** [review]
lgtm :tada:

**Review by borkopetrovicc — APPROVED** [review]

File: src/components/ProfileBox/AddNewProfile.tsx [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+type Props = {
+  onClick: () => void;
+};
+
+const AddNewProfile = ({ onClick }: Props) => {
+  const { t } = useTranslation('user');
+  const keyDownHandler = (event: React.KeyboardEvent<HTMLDivElement>) => (event.key === 'Enter' || event.key === ' ') && onClick();
+
+  return (
+    <div onClick={onClick} tabIndex={0} onKeyDown={keyDownHandler} className={classNames(styles.wrapper, styles.addProfileContainer)}>
```
Comment (ChristiaanScheermeijer) [inline src/components/ProfileBox/AddNewProfile.tsx:18]: `role="button"` is missing here 🤭

File: src/components/ProfileBox/ProfileBox.tsx [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+  onClick: () => void;
+  onEdit: () => void;
+  selected?: boolean;
+};
+
+const ProfileBox = ({ name, image, editMode = false, onClick, onEdit, selected = false }: Props) => {
+  return (
+    <div className={styles.wrapper}>
+      <div className={classNames(styles.inner, selected && styles.selected)}>
+        <div onClick={onClick} className={styles.box}>
```
Comment (ChristiaanScheermeijer) [inline src/components/ProfileBox/ProfileBox.tsx:25]: `role="button" tabIndex={0} onKeyDown={keyDownHandler}` is missing here for keyboard access (and a11y)

File: src/components/ProfileBox/ProfileBox.tsx [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+          <img className={styles.image} src={image || defaultAvatar} alt="" />
+        </div>
+        {editMode && (
+          <IconButton aria-label="edit-profile-button" onClick={onEdit} className={styles.overlay}>
```
Comment (ChristiaanScheermeijer) [inline src/components/ProfileBox/ProfileBox.tsx:29]: The `aria-label` should be translated and be human-readable text (without button because screen readers will announce the type already). E.g. `Edit profile`

File: src/components/UserMenu/ProfilesMenu/ProfilesMenu.tsx [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+  currentProfile,
+  small,
+  selectingProfile,
+  selectProfile,
+  defaultAvatar,
+  createButtonLabel,
+  switchProfilesLabel,
+  onCreateButtonClick,
+}: ProfilesMenuProps) => (
+  <>
```
Comment (ChristiaanScheermeijer) [inline src/components/UserMenu/ProfilesMenu/ProfilesMenu.tsx:35]: This component still creates invalid HTML markup using the LoadingOverlay and `hr` element inside an `ul`

Comment (ChristiaanScheermeijer) [reply src/components/UserMenu/ProfilesMenu/ProfilesMenu.tsx:35]: I see that the UserMenu also contains invalid HTML markup 🤭

File: src/components/UserMenu/UserMenu.tsx [hunk 20]
```diff
@@ -4,24 +4,36 @@ import { useTranslation } from 'react-i18next';
...
+  profilesEnabled?: boolean;
+  profiles?: Profile[];
 };
 
-const UserMenu = ({ showPaymentsItem, small = false, onClick }: Props) => {
+const UserMenu = ({ showPaymentsItem, small = false, onClick, accessModel, currentProfile, profilesEnabled, profiles }: Props) => {
   const { t } = useTranslation('user');
   const navigate = useNavigate();
 
+  const selectProfile = useSelectProfile();
```
Comment (ChristiaanScheermeijer) [inline src/components/UserMenu/UserMenu.tsx:35]: I think you forgot to move this hook to the container

File: src/containers/Profiles/EditProfile.tsx [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+            value: selectedAvatar || '',
+          }}
+          showCancelButton={!contained}
+          isMobile={isMobile}
+        />
+        <div className={styles.panel}>
+          <div className={styles.panelHeader}>
+            <h3>{t('profile.delete')}</h3>
+          </div>
+          <div className={profileStyles.profileInfo}>{t(profileDetails?.default ? 'profile.delete_main' : 'profile.delete_description')}</div>
```
Comment (ChristiaanScheermeijer) [inline src/containers/Profiles/EditProfile.tsx:100]: This isn't changed: https://github.com/jwplayer/ott-web-app/pull/353#discussion_r1314737204

File: src/containers/Profiles/Form.tsx [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    <form onSubmit={handleSubmit} noValidate>
+      <div className={styles.panel}>
+        <div className={styles.panelHeader}>
+          <h3>{formLabel}</h3>
+        </div>
+        <div className={profileStyles.profileInfo}>{t('profile.description')}</div>
+        <div className={profileStyles.formFields}>
+          {errors.form ? <FormFeedback variant="error">{errors.form}</FormFeedback> : null}
+          {submitting && <LoadingOverlay inline />}
+          <h2 className={profileStyles.nameHeading}>Name</h2>
```
Comment (ChristiaanScheermeijer) [inline src/containers/Profiles/Form.tsx:65]: Missing translation

File: src/containers/Profiles/Profiles.tsx [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+
+  const selectProfile = useSelectProfile();
+
+  if (loading || isLoading || isFetching) return <LoadingOverlay inline />;
+
+  return (
+    <div className={styles.wrapper}>
+      {activeProfiles === 0 ? (
+        <div className={styles.headings}>
+          <p className={styles.paragarph}>{t('profile.no_one_watching')}</p>
```
Comment (ChristiaanScheermeijer) [inline src/containers/Profiles/Profiles.tsx:49]: typo in className `paragraph`

File: src/containers/Profiles/Profiles.tsx [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+          />
+        ))}
+        {canAddNew && <AddNewProfile onClick={() => navigate('/u/profiles/create')} />}
+      </div>
+      {activeProfiles > 0 && (
+        <div className={styles.buttonContainer}>
+          {!editMode ? (
+            <Button onClick={() => navigate('/u/profiles/edit')} label={t('account.manage_profiles')} variant="outlined" size="large" fullWidth={isMobile} />
+          ) : (
+            <Button onClick={() => navigate('/u/profiles')} label="Done" variant="outlined" size="large" fullWidth={isMobile} />
```
Comment (ChristiaanScheermeijer) [inline src/containers/Profiles/Profiles.tsx:74]: Missing translation `label`

**Review by ChristiaanScheermeijer — COMMENTED** [review]
@naumovski-filip @kiremitrov123 I didn't look at the latest changes before the merge. I still have some feedback (no major issues).
